src/java/org/apache/cassandra/index/sai/disk/v1/postings/MergePostingList.java

```diff
@@ -98,18 +113,40 @@ else if (next == lastRowId)
     @Override
     public long advance(long targetRowID) throws IOException
     {
-        temp.clear();
-
-        for (var peekable : postingLists)
+        // clean out obsolete child lists, and remember the smallest row seen in case
+        // we can use it for the fast path
+        long nextRowId = PostingList.END_OF_STREAM;
+        PostingList nextPostingList = null;
+        var it = postingLists.iterator();
+        while (it.hasNext())
         {
+            var peekable = it.next();
             peekable.advanceWithoutConsuming(targetRowID);
-            if (peekable.peek() != PostingList.END_OF_STREAM)
-                temp.add(peekable);
+            long peeked = peekable.peek();
+            // clean out obsolete child lists
+            if (peeked == PostingList.END_OF_STREAM)
+                it.remove();
+
+            if (lastRowId <= peeked && peeked < nextRowId)
+            {
+                nextRowId = peeked;
+                nextPostingList = peekable;
+            }
+        }
+
+        if (lastRowId == targetRowID) {
+            // we're asking for the next row, past the current row, which an arbitrary
+            // number of our child posting lists may be pointing to.  In this case we
+            // need to let the PQ do its work to figure out what the correct next row AFTER
+            // this one is.
+            return nextPosting();
         }
-        postingLists.clear();
-        postingLists.addAll(temp);
 
-        return nextPosting();
+        // fast path -- no PQ
+        pq = null; // we're invalidating the pq's assumptions, so force a rebuild if we need it again
+        if (nextPostingList == null)
+            return PostingList.END_OF_STREAM;
+        return nextPostingList.nextPosting();
```

Comment:
Without the call to `nextPosting()`, I think we need something like:

```java
lastRowId = nextPostingList.nextPosting();
return lastRowId
```

Reply:
Ah, good catch.  Let me see if I can think of a good test for this.

Reply:
Turns out this is covered in the existing tests that I missed (and it was failing). Passes with that fix.